packages/protocol/test-sol/governance/validators/Validators.t.sol

```diff
@@ -0,0 +1,3204 @@
+      blsPublicKey,
+      blsPop
+    );
+  }
+}
+
+contract ValidatorsTest_DeregisterValidator_WhenAccountHasNeverBeenMemberOfValidatorGroup is
+  ValidatorsTest
+{
+  uint256 public constant INDEX = 0;
+
+  function setUp() public {
+    super.setUp();
+
+    _registerValidatorHelper(validator, validatorPk);
+
+    timeTravel(originalValidatorLockedGoldRequirements.duration);
+  }
+
+  function _deregisterValidator(address _validator) internal {
+    vm.prank(_validator);
+    validators.deregisterValidator(INDEX);
+  }
+
+  function test_ShouldMarkAccountAsNotValidator_WhenAccountHasNeverBeenMemberOfValidatorGroup()
+    public
+  {
+    assertTrue(validators.isValidator(validator));
+
+    _deregisterValidator(validator);
+
+    assertFalse(validators.isValidator(validator));
+  }
+
+  function test_ShouldRemoveAccountFromValidatorList_WhenAccountHasNeverBeenMemberOfValidatorGroup()
+    public
+  {
+    address[] memory ExpectedRegisteredValidators = new address[](0);
+
+    assertTrue(validators.isValidator(validator));
+    _deregisterValidator(validator);
+    assertEq(validators.getRegisteredValidators().length, ExpectedRegisteredValidators.length);
+  }
+
+  function test_ShouldResetAccountBalanceRequirements_WhenAccountHasNeverBeenMemberOfValidatorGroup()
+    public
+  {
+    assertTrue(validators.isValidator(validator));
+    _deregisterValidator(validator);
+    assertEq(validators.getAccountLockedGoldRequirement(validator), 0);
+  }
+
+  function test_Emits_ValidatorDeregisteredEvent_WhenAccountHasNeverBeenMemberOfValidatorGroup()
+    public
+  {
+    vm.expectEmit(true, true, true, true);
+    emit ValidatorDeregistered(validator);
+    _deregisterValidator(validator);
+  }
+
+  function test_Reverts_WhenAccountNotRegisteredValidator() public {
+    vm.expectRevert("Not a validator");
+    vm.prank(nonValidator);
+    validators.deregisterValidator(INDEX);
+  }
+
+  function test_Reverts_WhenWrongIndexProvided() public {
+    timeTravel(originalValidatorLockedGoldRequirements.duration);
+    vm.expectRevert("deleteElement: index out of range");
+    vm.prank(validator);
+    validators.deregisterValidator(INDEX + 1);
+  }
+}
+
+contract ValidatorsTest_DeregisterValidator_WhenAccountHasBeenMemberOfValidatorGroup is
+  ValidatorsTest
+{
+  uint256 public constant INDEX = 0;
+
+  function setUp() public {
+    super.setUp();
+
+    _registerValidatorHelper(validator, validatorPk);
+
+    _registerValidatorGroupHelper(group, 1);
+
+    vm.prank(validator);
+    validators.affiliate(group);
+
+    vm.prank(group);
+    validators.addFirstMember(validator, address(0), address(0));
+  }
+
+  function _deregisterValidator(address _validator) internal {
+    vm.prank(_validator);
+    validators.deregisterValidator(INDEX);
+  }
+
+  function test_ShouldMarkAccountAsNotValidator_WhenValidatorNoLongerMemberOfValidatorGroup()
+    public
+  {
+    vm.prank(group);
```

Comment:
Can we extract these lines into separate function so following tests can be shrinked?